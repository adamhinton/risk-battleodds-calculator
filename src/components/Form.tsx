import { useState } from "react";
import generateResults from "../utils/resultsCalculator";
import { UserInputs } from "../utils/resultsCalculator";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
//www.magicbell.com/blog/react-toast-notifications-made-easy
// I've tested this and it seems to work
const regEx = /^\s*\d+(\s*,\s*\d+)*\s*$/;

type FormProps = {
  setResults: Function;
};

type FormValues = {
  attackerCount: number;
  defenderCount: string;
  numSimulations: number;
};

const Form = (props: FormProps) => {
  const { setResults } = props;

  const [formValues, setFormValues] = useState<FormValues>({
    attackerCount: 10,
    defenderCount: "10",
    numSimulations: 1000,
  });

  function handleChange(evt: React.ChangeEvent<HTMLInputElement>) {
    const value = evt.target.value;
    console.log("evt.target.name:", evt.target.name);
    setFormValues({
      ...formValues,
      [evt.target.name]:
        evt.target.name === "defenderCount" ? value : Number(value),
    });
  }

  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();

        // Convert user's multiple defender inputs from a string to number[]
        const userInputs: UserInputs = {
          ...formValues,
          defenderCount: formValues.defenderCount.split(",").map((item) => {
            return Number(item);
          }),
        };

        const results = generateResults(userInputs);
        setResults(results);
      }}
    >
      <div>
        <label htmlFor="attackers">Attackers:</label>
        <input
          id="attackers"
          type="number"
          min="1"
          name="attackerCount"
          value={formValues.attackerCount}
          onChange={handleChange}
          data-testid="attackers-input"
        ></input>
      </div>

      <div>
        <label htmlFor="defenders">Defenders:</label>
        <input
          id="defenders"
          type="text"
          min="1"
          name="defenderCount"
          value={formValues.defenderCount}
          onChange={(e) => handleChange(e)}
          data-testid="defenders-input"
        ></input>
      </div>

      <div>
        <label htmlFor="simulations">Number of Simulations:</label>
        <input
          id="simulations"
          type="number"
          min="1"
          name="numSimulations"
          value={formValues.numSimulations}
          onChange={handleChange}
          data-testid="numsimulations-input"
        ></input>
      </div>

      <div>
        <button type="submit">Run Simulation</button>
      </div>
      <ToastContainer />
    </form>
  );
};

export default Form;
