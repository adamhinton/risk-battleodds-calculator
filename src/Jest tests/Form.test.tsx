import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import Form from "../components/Form";
const fakeSetState = jest.fn();

test("[1] renders without errors", () => {
  render(<Form setResults={fakeSetState} />);
});

test("[2] Renders three user input fields as expected", () => {
  render(<Form setResults={fakeSetState} />);
  const attackerInput = screen.getByTestId("attackers-input");
  const defenderInput = screen.getByTestId("defenders-input");
  const simulationsInput = screen.getByTestId("numsimulations-input");

  expect(attackerInput).toBeVisible();
  expect(defenderInput).toBeVisible();
  expect(simulationsInput).toBeVisible();

  expect(attackerInput).toHaveValue(10);
  expect(defenderInput).toHaveValue(10);
  expect(simulationsInput).toHaveValue(1000);
});

test("[3] Matches screenshot from 1.29.2023", () => {
  const tree = renderer.create(<Form setResults={fakeSetState} />).toJSON();
  expect(tree).toMatchSnapshot();
});
