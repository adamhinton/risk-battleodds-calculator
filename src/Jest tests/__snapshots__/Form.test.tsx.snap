// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[3] Matches screenshot from 1.29.2023 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
      htmlFor="attackers"
    >
      Attackers:
    </label>
    <input
      className="sc-beySbM iTRUCx"
      data-testid="attackers-input"
      id="attackers"
      name="attackerCount"
      onChange={[Function]}
      type="number"
      value={10}
    />
  </div>
  <div>
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
      htmlFor="defenders"
    >
      Defenders:
    </label>
    <input
      className="sc-beySbM iTRUCx"
      data-testid="defenders-input"
      id="defenders"
      name="defenderCount"
      onChange={[Function]}
      type="text"
      value="10"
    />
  </div>
  <div>
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-animated css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
      htmlFor="simulations"
    >
      Number of Simulations:
    </label>
    <input
      data-testid="numsimulations-input"
      id="simulations"
      min="1"
      name="numSimulations"
      onChange={[Function]}
      type="number"
      value={1000}
    />
  </div>
  <div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
      data-testid="submit-btn"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      Run Simulation
    </button>
  </div>
  <div
    className="Toastify"
  />
</form>
`;
